app/src/main/java/nl/example/kts/CoroutineViewModel.kt

```diff
@@ -14,8 +14,7 @@ abstract class CoroutineViewModel : ViewModel(),
     override val coroutineContext: CoroutineContext
         get() = Dispatchers.Main + job
 
-    override fun onCleared() {
+    final override fun onCleared() {
```

Comment:
This way, any additional clean up in `onCleared()` that might be desired by the user overriding this class will be impossible for him as he won't be able to override. I'd better mark this method with `@CallSuper` so the subclasses of this one when overriding this function will be forced to call super, thus the job will be canceled.

```
@CallSuper
override fun onCleared() {
    job.cancel()
}
```
This way, when a user overrides this class and overrides `onCleared()`, he will be forced to call `super.onCleared()` otherwise, he won't be able to compile.

Reply:
`@CallSuper` is not enforced at compile time IIRC, hence my suggestion to make it final.
There are multiple solutions if you want to allow receiving `onCleared` despite the method being final. Here are two examples:
- You can make the job a final protected property and allow implementers to register cleanup code with `invokeOnCancellation { … }`
- You can make an open `clear` or alike method that you call from the final `onCleared` implementation.

Reply:
Here is what happens if you don't call super:

<img width="438" alt="screen shot 2018-11-03 at 21 01 26" src="https://user-images.githubusercontent.com/1390865/47956859-0b321280-dfac-11e8-8aa6-cadf414f9211.png">

And here is what happens if you try to build from terminal:

<img width="1245" alt="screen shot 2018-11-03 at 21 03 33" src="https://user-images.githubusercontent.com/1390865/47956860-0b321280-dfac-11e8-920b-bdaa79a47da7.png">

Sure, there are tons of ways, but I don't think I have to complicate the things in this rather simple example where my main goal is to demonstrate something else. It's just a demo project :)

Reply:
Cool it now prevents compilation! `@CallSuper` alone is a good solution then.